File: src/207_Course_Schedule.py
# 207_Course_Schedule
# https://leetcode.cn/problems/course-schedule/


from typing import List

# 你这个学期必须选修 numCourses 门课程，记为 0 到 numCourses - 1 。

# 在选修某些课程之前需要一些先修课程。 先修课程按数组 prerequisites 给出，
# 其中 prerequisites[i] = [ai, bi] ，表示如果要学习课程 ai 则 必须 先学习课程  bi 。

# 例如，先修课程对 [0, 1] 表示：想要学习课程 0 ，你需要先完成课程 1 。
# 请你判断是否可能完成所有课程的学习？如果可以，返回 true ；否则，返回 false 。

# 示例 1：
# 输入：numCourses = 2, prerequisites = [[1,0]]
# 输出：true
# 解释：总共有 2 门课程。学习课程 1 之前，你需要完成课程 0 。这是可能的。

# 示例 2：
# 输入：numCourses = 2, prerequisites = [[1,0],[0,1]]
# 输出：false
# 解释：总共有 2 门课程。学习课程 1 之前，你需要先完成​课程 0 ；并且学习课程 0 之前，你还应先完成课程 1 。这是不可能的。

# 这个问题可以通过拓扑排序来解决。拓扑排序是一种对有向无环图（DAG）进行排序的算法。

# 首先，我们可以使用一个列表indegree来记录每门课程的入度（即有多少先修课程）。然后，我们使用一个邻接表adjacency来表示课程之间的依赖关系，
# 其中adjacency[i]表示课程i的后续课程列表。

# 接下来，我们从入度为0的课程开始，将其加入一个队列queue中。然后，我们不断从队列中取出课程，并处理其后续课程。
# 对于取出的课程a，我们遍历其后续课程列表adjacency[a]，将每门课程的入度减1。如果某门课程的入度减为0，则将其加入队列中。

# 最后，我们判断是否所有课程都被学习完。如果是，则返回true；否则，返回false。
from collections import deque

def canFinish(numCourses: int, prerequisites: List[List[int]]) -> bool:
    indegree = [0] * numCourses  # 记录每门课程的入度（有多少先修课程）
    adj = [[] for _ in range(numCourses)]  # 邻接表，表示课程之间的依赖关系
    
    for pair in prerequisites:
        course, prerequisite = pair
        indegree[course] += 1  # 更新课程的入度
        adj[prerequisite].append(course)  # 添加依赖关系
    
    queue = deque()  # 创建队列用于拓扑排序
    for i in range(numCourses):
        if indegree[i] == 0:
            queue.append(i)  # 将入度为0的课程加入队列
    
    while queue:
        course = queue.popleft()
        numCourses -= 1
        for next_course in adj[course]:
            indegree[next_course] -= 1  # 从取出的课程开始处理其后续课程
            if indegree[next_course] == 0:
                queue.append(next_course)  # 如果后续课程的入度为0，则加入队列
    
    return numCourses == 0  # 判断是否所有课程都被学习完
    
print(canFinish(numCourses = 2, prerequisites = [[1,0],[0,1]]))
    
    

File: src/207_Course_Schedule.rs
// 207_Course_Schedule
// https://leetcode.cn/problems/course-schedule/description/

// There are a total of numCourses courses you have to take, 
// labeled from 0 to numCourses - 1. 
// You are given an array prerequisites where prerequisites[i] = [ai, bi] indicates 
// that you must take course bi first if you want to take course ai.

// For example, the pair [0, 1], indicates that to take course 0 you have to first take course 1.
// Return true if you can finish all courses. Otherwise, return false.

// Example 1:
// Input: numCourses = 2, prerequisites = [[1,0]]
// Output: true
// Explanation: There are a total of 2 courses to take. 
// To take course 1 you should have finished course 0. So it is possible.

// Example 2:
// Input: numCourses = 2, prerequisites = [[1,0],[0,1]]
// Output: false
// Explanation: There are a total of 2 courses to take. 
// To take course 1 you should have finished course 0, an

use std::collections::HashMap;
use std::collections::VecDeque;

impl Solution {
    pub fn can_finish(num_courses: i32, prerequisites: Vec<Vec<i32>>) -> bool {
        // prerequisites_count 用于存储每门课程所需的先修课程数量。
        let mut prerequisites_count = vec![0; num_courses as usize];
        
        // course_map 映射了一门课程到一列依赖于这门课程的后续课程。
        let mut course_map = HashMap::new();
        
        // 根据给定的先修课程关系，填充 prerequisites_count 和 course_map。
        for pair in &prerequisites {
            let (course, prerequisite) = (pair[0] as usize, pair[1] as usize);
            prerequisites_count[course] += 1;
            course_map.entry(prerequisite).or_insert(vec![]).push(course);
        }

        // 队列用于存放没有先修课程限制的课程。
        let mut queue = VecDeque::new();
        
        // 将没有先修课程的课程添加到队列中。
        for (index, &count) in prerequisites_count.iter().enumerate() {
            if count == 0 {
                queue.push_back(index);
            }
        }

        // 记录可以选的课程数量。
        let mut taken_courses = 0;

        // 如果当前课程没有先修课程限制，就将其从队列中删除，并为取出的课程增加已选课程计数。
        while let Some(course) = queue.pop_front() {
            taken_courses += 1;
            
            // 对于每个依赖于当前课程的后续课程，减少其 prerequisites_count。
            // 如果某后续课程的 prerequisites_count 减少到 0，则将其加入队列。
            if let Some(subsequent_courses) = course_map.get(&course) {
                for &subsequent_course in subsequent_courses {
                    prerequisites_count[subsequent_course] -= 1;
                    if prerequisites_count[subsequent_course] == 0 {
                        queue.push_back(subsequent_course);
                    }
                }
            }
        }

        // 如果我们可以选所有的课程，返回 true，否则返回 false。
        taken_courses == num_courses
    }
}
